operator/apis/loki/v1/v1.go

```diff
@@ -83,6 +83,8 @@ var (
 	ErrReplicationSpecConflict = errors.New("replicationSpec and replicationFactor (deprecated) cannot be used at the same time")
 	// ErrIPv6InstanceAddrTypeNotAllowed when the default InstanceAddrType is used with enableIPv6.
 	ErrIPv6InstanceAddrTypeNotAllowed = errors.New(`instanceAddrType "default" cannot be used with enableIPv6 at the same time`)
+	// ErrDesiredRateAndPerStreamRateNotAllowed when there's a conflict in the rate limiting configuration because both desiredRate and perStreamRateLimit are set.
+	ErrDesiredRateAndPerStreamRateNotAllowed = errors.New("fields desiredRate and perStreamRateLimit cannot be used together")
```

Comment:
fyi: in some scenarios you'd benefit by tuning both. Ex: if you ingest a spike going from 0MB/s ingestion to 5MB/s, you'll hit the per-stream rate limit even with stream sharding enabled, because Loki wouldn't have enough time to adjust to the current load. In this scenario, setting the perStreamRateLimit to 5MB/s would help avoiding discards.

Reply:
I'm not sure I understand this correctly. I was under the impression, that using `desiredRate` without setting a `perStreamRateLimit` would effectively turn the per-stream rate-limiting into a "soft limit" where peaks would not return a 429 and having a stream with high log rate would be turned into multiple streams internally automatically to avoid having a single large stream.

Reply:
it is a little confusing. Because it takes time to propagate the right number of shards, if you get a spike too quickly for a stream the system won't have enough time to adjust. In this scenario, the per-stream-rate-limit dictates if your logs will be rejected or not.

Reply:
FWIW and IIF I am reading the rate store code correctly, the rates are based on observed data points on the ingester side. So for things not seen yet we cannot create adapt the shard count. Thus spiky streams will be rate limited by the per-stream rate limit. Correct?

Reply:
I am referring to this: https://github.com/grafana/loki/blob/v2.9.2/pkg/distributor/ratestore.go#L301

Reply:
> FWIW and IIF I am reading the rate store code correctly, the rates are based on observed data points on the ingester side. So for things not seen yet we cannot create adapt the shard count. Thus spiky streams will be rate limited by the per-stream rate limit. Correct?

exactly it.

Reply:
> FWIW and IIF 

just learned what FWIW means, but couldn't really find what IIF stands for :rofl:

Reply:
The emoji and acronym domination among engineering folks is sometimes excessive. sorry for that :see_no_evil:

Reply:
nrta, idimsf

(no reason to apologize, I do it myself)

Reply:
cool, I'll resolve the thread. if you need any clarifications regarding stream sharding please let me know (you can ping me on Slack)